src/librbd/api/Image.cc

```diff
@@ -601,10 +601,24 @@ int Image<I>::remove(IoCtx& io_ctx, const std::string &image_name,
     ConfigProxy config(cct->_conf);
     Config<I>::apply_pool_overrides(io_ctx, &config);
 
+    std::string get_image_id;
+    int r = cls_client::dir_get_id(&io_ctx, RBD_DIRECTORY, image_name,
+                                   &get_image_id);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "failed to retrieve image id: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    int ret = Trash<I>::move(
+      io_ctx, RBD_TRASH_IMAGE_SOURCE_USER, image_name,
+      config.get_val<uint64_t>("rbd_move_to_trash_on_remove_expire_seconds"));
+
     if (config.get_val<bool>("rbd_move_to_trash_on_remove")) {
-      return Trash<I>::move(
-        io_ctx, RBD_TRASH_IMAGE_SOURCE_USER, image_name,
-        config.get_val<uint64_t>("rbd_move_to_trash_on_remove_expire_seconds"));
+      return ret;
```

Comment:
I think you should also return if `Trash::move` returned an error. Or may be proceed with `librbd::image::RemoveRequest` in case the trash move has failed due to a corrupted image. May be proceed only if `-ENOENT` returned? But I suppose this could be due to a simultaneous removal race, and in such a case it is better just to return.

@dillaman what do you think?

Reply:
It would be great if we can solve the race condition of two folks running "rbd rm XYZ" and both start moving the image to the trash simultaneously. I don't think it should fall back to removing the image silently in such cases.

Reply:
@dillaman the difference betwee `Trash::move->Trash::remove` and `librbd::image::RemoveRequest` is that the later is tolerant to missed metadata. So usually you can remove a damaged image with it (at least it will try its best) or continue an interrupted removal. With just "trash move, trash remove" it looks like we will lost this functionality.

Reply:
@trociny What missing metadata are you thinking of in this case? Like missing the full trash directory entry?

Reply:
I mean, the first step that `Trash::move` does it opens the image. If the open fails, the trash move fails too. So if we just abort after unsuccessful trash move, the remove operation will fail.  In the case of `RemoveRequest`, it also tries to open the image first, but if it fails it still continue with removing what it can.

For example right now, if I remove rbd_header object for an image, `rbd rm` is still able to remove the image entry from directory and its rbd_id object (although it can't remove data objects). With "trash move -> trash remove" it will not be able to do this.

The removed "rbd_header" is just an example. I suppose there are many ways to corrupt image metadata so the open will fail.

BTW, in the case  of `RemoveRequest`, it disables journal before acquiring lock, in case it is corrupted.

Reply:
I think for simultaneous execution of `rbd rm XYZ`, we could perhaps return  on`Trash::move' failure. That way, for instance if it fails on obtaining exclusive lock for one user, it may succeed for the other. But this won't foolproof other failure scenarios.

Reply:
The "trash move" already uses a state enum to track in-progress actions. However, if a "trash move" dies between starting the action and finishing, it will leave it in an inconsistent state so we need to be able to re-run "trash move" (and therefore it would also support simultaneous execution). 

Perhaps "trash move" just needs to behave more like "remove" in that it doesn't actually open the image, it just checks that it doesn't have watchers (besides itself) and proceeds to update the directory and remove the id object.

Reply:
The ```RemoveRequest``` state machine does attempt to open the image, but that's because it actually has to do some work and potentially avoid corruption of in-use images, or leaking snapshots, etc. In the case of ```Trash::move```, since we don't touch any of the underlying image data, it should always be safe to just add it to the trash dictionary, remove the id object, and remove it from the image directory. However, adding a check to see that there are no other watches on the image at least provides a sanity check.